Rename AcitivityReporter and auto-create publishing steps

captainsafia opened the pull request:
### Summary

This PR significantly refactors the publishing and pipeline architecture to simplify the activity reporting model. The key change is renaming `DeployingContext` to `PipelineContext` and consolidating the activity reporter hierarchy from a three-level step/task structure to a simpler two-level structure where pipeline steps directly manage tasks.

### Changes Overview

**Core API Renaming:**
- `DeployingContext` → `PipelineContext`
- PipelineStepContext (new)
- `IPublishingActivityReporter` → `IPipelineActivityReporter`
- `PublishingActivityReporter` → `PipelineActivityReporter`
- `NullPublishingActivityReporter` → `NullPipelineActivityReporter`
- `IPublishingStep` → `IReportingStep`
- - `IPublishingTask` → `IReportingTask`

**Architecture Simplification:**

The previous three-level hierarchy:
```
ActivityReporter → Step → Task
```

Has been simplified to integrate step management directly into the pipeline execution:
```
PipelineContext.Publishing → Task
```

**Pipeline Execution Changes:**

Steps are now created automatically during pipeline execution rather than requiring explicit creation within step actions. Before:

```csharp
builder.Pipeline.AddStep("assign-storage-role", async (context) =>
{
    var roleAssignmentStep = await context.ActivityReporter
        .CreateStepAsync($"assign-storage-role", context.CancellationToken);
    
    await using (roleAssignmentStep.ConfigureAwait(false))
    {
        var assignRoleTask = await roleAssignmentStep
            .CreateTaskAsync($"Granting file share access...", context.CancellationToken);
        // ... task work
    }
});
```

After:
```csharp
builder.Pipeline.AddStep("assign-storage-role", async (deployingContext) =>
{
    var assignRoleTask = await deployingContext.ReportingStep
        .CreateTaskAsync($"Granting file share access...", deployingContext.CancellationToken);
    
    await using (assignRoleTask.ConfigureAwait(false))
    {
        // ... task work
    }
});
```

captainsafia committed:
Rename AcitivityReporter and auto-create publishing steps

Rename deploying to pipeline in more places

Rename type to PublishingStep

davidfowl commented on playground/pipelines/Pipelines.AppHost/AppHost.cs:
```diff
@@ -46,9 +46,9 @@
 // This step could also be modeled as a Bicep resource with the role assignment
 // for the principalId associated with the deployment.
-builder.Pipeline.AddStep("assign-storage-role", async (context) =>
+builder.Pipeline.AddStep("assign-storage-role", async (deployingContext) =>
```
context

davidfowl reviewed (changes requested):
PipelineContext can't have PublishingStep on it, it gets passed to many steps concurrently

davidfowl committed:
Refactor pipeline context to use PipelineStepContext for step-specific execution

davidfowl committed:
Refactor publishing-related code to use pipelines

- Removed references to Aspire.Hosting.Publishing in multiple files and replaced them with Aspire.Hosting.Pipelines.
- Introduced new classes: CompletionState, IPipelineActivityReporter, IPublishingStep, IPublishingTask, and their implementations.
- Created a NullPublishingActivityReporter for testing purposes.
- Added PipelineActivityReporter to manage publishing activities.
- Implemented PipelineContext and PipelineStepContext for better context management during pipeline execution.
- Updated tests to reflect changes in namespaces and class references.

davidfowl committed:
Refactor publishing components to use reporting interfaces and improve task management

davidfowl committed:
Refactor pipeline context to replace PublishingStep with ReportingStep for improved task management

davidfowl committed:
Refactor tests to replace PublishingStep with ReportingStep and update related interfaces

davidfowl committed:
Add compatibility suppressions for new publishing-related types

davidfowl reviewed (approved):
LGTM, but waiting for extra @captainsafia review

captainsafia committed:
Fix sample parameter name

captainsafia commented:
LGTM overall!

Nit about API names: `PipelineContext` and `PipelineStepContext` don't make it clear what stage of the pipeline flow you are in. Taking inspiration from endpoint filters, maybe we do something like `PipelineFactoryContext` and `PipelineStepInvocationContext`? Although I hate class names that are that long. Food for thought...
